File: hw/palomo/u3/hw17OOP/Phone.TypeScript/ec.edu.espe.model/CellPhone.ts
import { Device } from "./Device";
import { Battery } from "./Battery";
import { Screen } from "./Screen";
import { SIMCard } from "./SIMCard";

export class CellPhone extends Device {

    constructor(
        brand: string,
        model: string,
        private imei: string,

        // Composition
        private battery: Battery,

        // Association
        private screen: Screen,

        // Aggregation
        private simCard: SIMCard
    ) {
        super(brand, model);
    }

    override turnOn(): void {
        console.log("Phone ON");
    }

    override turnOff(): void {
        console.log("Phone OFF");
    }

    public makeCall(number: string): void {
        console.log(`Calling ${number}`);
    }

    public showInfo(): void {
        console.log("\nPHONE INFORMATION");
        console.log(`Brand: ${this.getBrand()}`);
        console.log(`Model: ${this.getModel()}`);
        console.log(`IMEI: ${this.imei}`);
        console.log(`Battery: ${this.battery.getCapacity()} mAh`);
        console.log(`Screen: ${this.screen.getSize()} inches`);
        console.log(`Operator: ${this.simCard.getOperator()}`);
    }

    public toJSON() {
        return {
            brand: this.getBrand(),
            model: this.getModel(),
            imei: this.imei,
            battery: this.battery.getCapacity(),
            screen: this.screen.getSize(),
            operator: this.simCard.getOperator()
        };
    }
}

File: hw/palomo/u3/hw17OOP/cell_phone_project/ec_edu_espe_model/cell_phone.py
from ec_edu_espe_model.device import Device

class CellPhone(Device):

    def __init__(self, brand, model, imei,
                 battery, screen, sim_card):

        super().__init__(brand, model)

        self._imei = imei

        # Composition
        self._battery = battery

        # Association
        self._screen = screen

        # Aggregation
        self._sim_card = sim_card

    def turn_on(self):
        print("Phone ON")

    def turn_off(self):
        print("Phone OFF")

    def make_call(self, number):
        print(f"Calling {number}")

    def to_dict(self):

        return {
            "brand": self.brand,
            "model": self.model,
            "imei": self._imei,
            "battery": self._battery.capacity,
            "screen": self._screen.size,
            "operator": self._sim_card.operator
        }

    def show_info(self):

        print("\nPHONE INFORMATION")
        print(f"Brand: {self.brand}")
        print(f"Model: {self.model}")
        print(f"IMEI: {self._imei}")
        print(f"Battery: {self._battery.capacity} mAh")
        print(f"Screen: {self._screen.size}\"")
        print(f"Operator: {self._sim_card.operator}")